### ISSUE TITLE
Improve error message of generic declarations using [_] as type, as `undeclared identifier _` compiler error is hard to understand

### ISSUE THREAD

**Issue by PhilippMDoerner** [nim-lang/Nim#21435]
### Description

It used to be accepted to have generic declarations that made use of `_`.

[An example of this is norm](https://github.com/moigagoo/norm/blob/4479240e6a3347fa8990839b7240d346c4b686a6/src/norm/private/sqlite/dbtypes.nim#L49) that currently runs on nim 1.6.10 with generics like this (And this did work with the nim devel version from 20.12.2022) :

```nim
func dbValue*[_](val: StringOfCap[_]): DbValue = dbValue(string(val))

func dbValue*[_](val: PaddedStringOfCap[_]): DbValue = dbValue(string(val))
```

Now a nim file with this one line of code will not compile:

```nim
proc wontCompile[_](x, y: _) = echo "Yay"
```

### Nim Version

Nim Compiler Version 1.9.1 [Linux: amd64]
Compiled at 2023-02-24
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: 7dc80a7fefb0d2cecbdc2696c71ea2fa94d6aa76
active boot switches: -d:release

### Current Output

```text
Error: undeclared identifier: '_'
candidates (edit distance, scope distance); see '--spellSuggest': 
 (1, 3): '$'
 (1, 3): '&'
 (1, 3): '*'
 (1, 3): '+'
 (1, 3): '-'
 (1, 3): '/'
 (1, 3): '<'
 (1, 3): '='
 (1, 3): '>'
 (1, 3): '@'
 (1, 3): '^'
 (1, 3): '|'
 (2, 3): '!='
 (2, 3): '%%'
 (2, 3): '&='
 (2, 3): '*%'
 (2, 3): '*='
 (2, 3): '+%'
 (2, 3): '+='
 (2, 3): '-%'
 (2, 3): '-='
 (2, 3): '..'
 (2, 3): '/%'
 (2, 3): '/='
 (2, 3): '<%'
 (2, 3): '<='
 (2, 3): '=='
 (2, 3): '>%'
 (2, 3): '>='
 (2, 3): '[]'
 (2, 3): 'in'
 (2, 3): 'is'
 (2, 3): 'of'
 (2, 3): 'on'
 (2, 3): 'or'
 (2, 3): 'ze'
 (2, 3): '||'
 (3, 3): '..<'
 (3, 3): '..^'
 (3, 3): '<=%'
 (3, 3): '>=%'
 (3, 3): 'Inf'
 (3, 3): 'NaN'
 (3, 3): '[]='
 (3, 3): 'abs'
 (3, 3): 'add'
 (3, 3): 'and'
 (3, 3): 'any'
 (3, 3): 'cas'
 (3, 3): 'chr'
```

### Expected Output

```text
No error
```

### Possible Solution

No clue

### Additional Information

After chatting with @ringabout the culprit that introduced this bug (regression?) appears to have been https://github.com/nim-lang/Nim/pull/21192 and they stated they were able to confirm it was the culprit.

Is this maybe intended behaviour that I'm missing because of my lack of understanding?
@metagn as they likely have a better understanding of the matter than I do.

Would be good to know if this behaviour will get changed in the future or not, as I'll need to update norm`s codebase if this behaviour remains. Which isn't a big deal, I'd just like to get a headsup and prefer not to touch the norm code there unless necessary :smile:

**Comment by metagn** [comment]
In #21192 the ability to add anything with the name `_` to scope was removed ([this line](https://github.com/nim-lang/Nim/pull/21192/files#diff-7fb8657a09bc02fa86de64a0ee8cba5392d40798dd0f6523f66195b02e069430R326)). I don't know why I didn't mention this when I made the PR and made it seem like it was just about routine parameters. For example `type _ = int; var x: _` also works in 1.6 but doesn't work in devel.

I'm guessing this was to not allow situations to slip through where `_` can be used when it shouldn't be, in the specific case of routine parameters they can be added to scope in 1. proc bodies 2. template bodies 3. proc types (for the type of default parameters) and maybe more that I didn't know about.

We can revert this but it would mean we have to specifically find all situations where we want `_` to not be usable and add `if name != "_"` (or the reverse and `addDeclAt(allowUnderscore = true)` etc). My main problem with this is the resulting list would probably be too subjective/the logic would be hard to reason about. For generic parameters I can kind of see it but it could still be convenient to have like `proc foo[_, _]` e.g. in overloading. The logic of "_ will never be a symbol in scope" is much more simple and useful IMO.

In this case, I guess it sucks to have to change the code but it's not really a big change, one can rename `_` to `T` or remove the `[_]` entirely and use implicit generics like `func dbValue*(val: StringOfCap)`. If it really matters though I can still make the PR to allow generic parameters named `_`. Decision is not up to me.

**Comment by PhilippMDoerner** [comment]
I'm fine with this as well, what I'm mostly wanting to know is what is the *intended* way this should work? What is the desired specification? 

If it is that `[_]` should work, then that is nice for me because then this is actually a bug and should be fixed.

If it is that `[_]` shouldn't work then this is a breaking change of nim2.0 and something we need to address in norm. Which is perfectly fine, I'd just like to know that it's necessary.

Is this the kind of question that requires Araq for an ultimate statement about intended specs?
Given how seemingly menial it is I'm hesitant. Particularly given that @metagn  's explanation makes perfect sense to me, so I'm leaning towards just declaring this as working as intended. 

Maybe improve the error message there could be done, because for this kind of issue just `Error: undeclared identifier: '_'` is a fairly nondescript message that a user wouldn't know what it means (heck, I wouldn't). 

@ringabout you're pretty much the first instance I go to in these kinds of scenarios, do you have an opinion on this?

**Comment by beef331** [comment]
I'd say the best solution is inside undeclared identifier specialize an error message for `_` that states something along the lines of "`_` is a special identifier that does not introduce a symbol" instead of the present error for it. `_` should never be used if you actually want to use the identifier.

**Comment by ringabout** [comment]
Yeah, I agree. And it is better to document it somewhere like in the changelog.

**Comment by PhilippMDoerner** [comment]
> Yeah, I agree. And it is better to document it somewhere like in the changelog.

PR with mention of it in the docs has been created.
I intentionally did not add the comment that it closes #21435 , because it shouldn't.

I'd consider the task of this issue now to change the error message that occurs when somebody writes `proc foo[_](x: typedesc[_]): string = "BAR"` to something more understandable.

Should I change the initial description and heading of this issue to clarify that?

**Comment by Araq** [comment]
> What is the desired specification?

As others have also said, it's "name here that cannot be accessed later on". Regardless of the context.
